firestore/functions/src/index.ts

```diff
@@ -0,0 +1,111 @@
+ */
+export const helloWorld = functions.https.onRequest((request, response) => {
+ response.send("Hello from Firebase!");
+});
+
+/**
+ * Callable cloud function to get back a list of favorite restaurants for a user
+ */
+export const getFavorites_v0 = functions.https.onCall(async (data, context) => {
+    const userID = data.uid;
+    try {
+        if (userID) {
+            console.log(`I am going to get favorites for user ${userID}`);
+            const userDoc = await admin.firestore().doc(`users/${userID}`).get();
+            console.log(`Here is my user doc`);
+            console.log(util.inspect(userDoc.data()));
+            const userData = userDoc.data();
+            if (!userData) {
+                console.error('Data not found for ', userID);
+                return([]);
+            }
+            const favorites = userData.favorites;
+            if (!favorites) {
+                console.log('User has no favorites');
+                return([]);
+            }
+            const fetchPromises: Promise<DocumentSnapshot>[] = [];
+            favorites.forEach((restID: string)  => {
+                console.log('I\'m gonna fetch data for ', restID);
+                const nextPromise = admin.firestore().doc(`restaurants/${restID}`).get();
+                fetchPromises.push(nextPromise);
+            });
+            return Promise.all(fetchPromises).then((docSnaps) => {            
+                const responseArray: DocumentData[] = []
+                docSnaps.forEach((snapshot: DocumentSnapshot) => {
+                    const docData = snapshot.data()!;
+                    docData['.id'] = snapshot.id;
+                    responseArray.push(docData);
+                });
+                return(responseArray);
+            });
+        } else {
+            console.error('No user ID');
+            return([]);
+        }
+    } catch (error) {
+        console.error('Received an error: ', error)
+        throw new functions.https.HttpsError('internal', 'Got some unknown error ' + error);
+    }
+});
+
+/**
+ * The getFavorites function as an HTTPS call, which was a little easier to demo at first
+ */
+export const getFavorites_as_http = functions.https.onRequest(async (request, response) => {
+    const userID = request.query.uid;
+    if (userID !== null) {
+        console.log(`I am going to get favorites for user ${userID} <br/>`);
+        const userDoc = await admin.firestore().doc(`users/${userID}`).get();
+        console.log(`Here is my user doc <br/>`);
+        console.log(util.inspect(userDoc.data()));
+        const userData = userDoc.data();
+        if (userData === null) {
+            response.send('No user data');
+            return
+        }
+        const favorites = userData!.favorites;
+        if (favorites === null) {
+            response.send('You have no favorites');
+            return
+        }
+        const fetchPromises: Promise<DocumentSnapshot>[] = [];
```

Comment:
This is bit of a stylistic preference, but a `map` works nicely here. It also means you don't have to explicitly type the array.

```ts
const fetchPromises  = favorites.map(restID => {
  console.log('I\'m gonna fetch data for ', restID);
  return admin.firestore().doc(`restaurants/${restID}`).get();
});
```

You can even put it on one line if you want to omit the `console.log`:

```ts
const firestore = admin.firestore();
const fetchPromises  = favorites.map(id => firestore().doc(`restaurants/${id}`).get());
```

Reply:
Done, although TypeScript still got angry at me for not adding an explicit type to the array.